samples/subsys/usb/hid-mouse/src/main.c

```diff
@@ -92,75 +149,178 @@ static const u8_t hid_report_desc[] = {
 		/* END_COLLECTION				C0    */
 };
 
-static int get_report_cb(struct usb_setup_packet *setup, s32_t *len,
-			 u8_t **data)
-{
-	*len = sizeof(hid_report_desc);
-	*data = (u8_t *)hid_report_desc;
+static struct hid_ops ops;
 
-	return 0;
-}
+static u32_t def_val[4];
+static volatile u8_t status[4];
+static K_SEM_DEFINE(sem, 0, 1);	/* starts off "not available" */
+static struct gpio_callback callback[4];
 
-static struct hid_ops ops = {
-	.get_report = get_report_cb,
-};
+#define MOUSE_BTN_REPORT_POS	0
+#define MOUSE_X_REPORT_POS	1
+#define MOUSE_Y_REPORT_POS	2
 
-static int def_val;
-static volatile int status;
-static K_SEM_DEFINE(sem, 0, 1);	/* starts off "not available" */
+#define MOUSE_BTN_LEFT		BIT(0)
+#define MOUSE_BTN_RIGHT		BIT(1)
+#define MOUSE_BTN_MIDDLE	BIT(2)
 
-#define MOUSE_BTN_LEFT			BIT(0)
-#define MOUSE_BTN_RIGHT			BIT(1)
-#define MOUSE_BTN_MIDDLE		BIT(2)
 
-void button_pressed(struct device *gpio, struct gpio_callback *cb,
-		    u32_t pins)
+static void left_button(struct device *gpio, struct gpio_callback *cb,
+		 u32_t pins)
 {
-	int cur_val, state = status;
+	u32_t cur_val;
+	u8_t state = status[MOUSE_BTN_REPORT_POS];
 
-	gpio_pin_read(gpio, PIN, &cur_val);
-	if (def_val != cur_val) {
+	gpio_pin_read(gpio, PIN0, &cur_val);
+	if (def_val[0] != cur_val) {
 		state |= MOUSE_BTN_LEFT;
 	} else {
 		state &= ~MOUSE_BTN_LEFT;
 	}
 
-	if (state != status) {
-		status = state;
+	if (status[MOUSE_BTN_REPORT_POS] != state) {
+		status[MOUSE_BTN_REPORT_POS] = state;
+		k_sem_give(&sem);
+	}
+}
+
+#ifdef SW1_GPIO_PIN
+static void right_button(struct device *gpio, struct gpio_callback *cb,
+		  u32_t pins)
+{
+	u32_t cur_val;
+	u8_t state = status[MOUSE_BTN_REPORT_POS];
+
+	gpio_pin_read(gpio, PIN1, &cur_val);
+	if (def_val[0] != cur_val) {
+		state |= MOUSE_BTN_RIGHT;
+	} else {
+		state &= ~MOUSE_BTN_RIGHT;
+	}
+
+	if (status[MOUSE_BTN_REPORT_POS] != state) {
+		status[MOUSE_BTN_REPORT_POS] = state;
 		k_sem_give(&sem);
 	}
 }
+#endif
+
+#ifdef SW2_GPIO_PIN
+static void x_move(struct device *gpio, struct gpio_callback *cb,
+	    u32_t pins)
+{
+	u32_t cur_val;
+	u8_t state = status[MOUSE_X_REPORT_POS];
+
+	gpio_pin_read(gpio, PIN2, &cur_val);
+
+	if (def_val[2] != cur_val) {
+		state += 10;
+	} else {
+
+	}
+
+	if (status[MOUSE_X_REPORT_POS] != state) {
+		status[MOUSE_X_REPORT_POS] = state;
+		k_sem_give(&sem);
+	}
+}
+#endif
+
+#ifdef SW3_GPIO_PIN
+static void y_move(struct device *gpio, struct gpio_callback *cb,
+	    u32_t pins)
+{
+	u32_t cur_val;
+	u8_t state = status[MOUSE_Y_REPORT_POS];
+
+	gpio_pin_read(gpio, PIN3, &cur_val);
+
+	if (def_val[3] != cur_val) {
+		state += 10;
+	} else {
+
+	}
+
+	if (status[MOUSE_Y_REPORT_POS] != state) {
+		status[MOUSE_Y_REPORT_POS] = state;
+		k_sem_give(&sem);
+	}
+}
+#endif
+
+int configure_callback(struct device *gpio, u32_t pin, int flags,
```

Comment:
Is this a callback?

Reply:
No, it configures GPIOs and adds callbacks to button pins. I see now this might not be the best name. `callbacks_configure()` maybe?

Reply:
Yes, this sound better, at first it sounds like a callback